### ISSUE TITLE
Get-BcContainerIpAddress is not always returning an IP address

### ISSUE THREAD

**Issue by BillJay1821** [microsoft/navcontainerhelper#2970]
BcContainerHelper 4.0.15.
BcContainerHelper is running as administrator in PowerShell ISE.

```
Name                           Value                                                                                                                                                                                                                              
----                           -----                                                                                                                                                                                                                              
PSVersion                      5.1.17763.3770                                                                                                                                                                                                                     
PSEdition                      Desktop                                                                                                                                                                                                                            
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                                            
BuildVersion                   10.0.17763.3770                                                                                                                                                                                                                    
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                                    
WSManStackVersion              3.0                                                                                                                                                                                                                                
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                                
SerializationVersion           1.1.0.1   
```

When compiling an AL folder under normal conditions there's a line in the output as seen below.

```
Downloading symbols: Microsoft_Application_22.0.53722.0.app
Url : http://172.31.38.43:7049/BC/dev/packages?publisher=Microsoft&appName=Application&versionText=21.5.0.0&tenant=default
```
However, I'm often seeing the line below which causes Compile-AppInNavContainer to fail.

```
Downloading symbols: Microsoft_Application_22.0.54157.54941.app
Url : http://MSFT_NetIPAddress (Name = ";?A8:8:8;55?55;55;", CreationClassName = "", SystemCreationClassName = "", SystemName = ""):7049/BC/dev/packages?publisher=Microsoft&appName=Application&versionText=21.5.0.0&tenant=default
```
I traced this to the line in Compile-AppInNavContainer.

`$ip = Get-BcContainerIpAddress -containerName $containerName`

When Get-BcContainerIpAddress is working properly, it returns a regular IP address.

```
C:\Get-BcContainerIpAddress -containerName $containerName
172.31.38.43
```

When Get-BcContainerIpAddress is NOT working properly, it returns an object as seen below.

```
ifIndex                  : 3
PSComputerName           : 249cceb2d487e70e6fffb1f797495efd17e8fd7da27f5d4f44bbf25658f217f9
RunspaceId               : 8c70375e-922d-413e-98e0-6b7d3047725c
Caption                  : 
Description              : 
ElementName              : 
InstanceID               : 
CommunicationStatus      : 
DetailedStatus           : 
HealthState              : 
InstallDate              : 
Name                     : ;<A8:8:8;55=55;55;
OperatingStatus          : 
OperationalStatus        : 
PrimaryStatus            : 
Status                   : 
StatusDescriptions       : 
AvailableRequestedStates : 
EnabledDefault           : 2
EnabledState             : 
OtherEnabledState        : 
RequestedState           : 12
TimeOfLastStateChange    : 
TransitioningToState     : 12
CreationClassName        : 
SystemCreationClassName  : 
SystemName               : 
NameFormat               : 
OtherTypeDescription     : 
ProtocolIFType           : 4096
ProtocolType             : 
Address                  : 
AddressOrigin            : 0
AddressType              : 
IPv4Address              : 127.0.0.1
IPv6Address              : 
IPVersionSupport         : 
PrefixLength             : 8
SubnetMask               : 
AddressFamily            : 2
AddressState             : 4
InterfaceAlias           : Loopback Pseudo-Interface 2
InterfaceIndex           : 3
IPAddress                : 127.0.0.1
PreferredLifetime        : 10675199.02:48:05.4775807
PrefixOrigin             : 2
SkipAsSource             : False
Store                    : 1
SuffixOrigin             : 2
Type                     : 1
ValidLifetime            : 10675199.02:48:05.4775807
```
What's odd is that although it is happening most of the time, it is not 100% of the time.

For example, in the most recent round of testing, it successfully returned the IP address for the first application that compiled and failed on the second application. When I recompile the first app it is now failing again and Get-BcContainerIpAddress only returns the object.

I am creating the BC container the same way I always have, so nothing has changed there.

The only thing that changed was last week I was using BcContainerHelper 4.0.14.

I tried reverting from BcContainerHelper 4.0.15 to 4.0.14 and it made no difference, so I can't really say it is something in the version, even though 4.0.14 was working last week.

I've also tried multiple docker containers at this point.

I just noticed that my docker version is older. I'm going to try updating docker.

```
Latest stable available Docker Engine version is 23.0.3
Current installed Docker Engine version 20.10.22
```

Get-BcContainerIpAddress worked once after updating docker, but started failing again.

This is the code fragment in Get-BcContainerIpAddress that is failing.

```
    $ip = Invoke-ScriptInBcContainer -containerName $containerName -scriptblock {
        if ($global:IP) {
            $ip = $global:IP
        }
        else {
            $ip = ""
            $ips = Get-NetIPAddress | Where-Object { $_.AddressFamily -eq "IPv4" -and $_.IPAddress -ne "127.0.0.1" }
            if ($ips) {
                $ips | ForEach-Object {
                    if ("$ip" -eq "") {
                        $ip = $_.IPAddress
                    }
                }
            }
            $global:IP = $ip
        }
        $ip
    }

Write-Output ("[" + $ip + "]")
```

and produces this output

`[MSFT_NetIPAddress (Name = ";?A8:8:8;55?55;55;", CreationClassName = "", SystemCreationClassName = "", SystemName = "")]
`

This fragment of the code will return a proper IP address when executed inside the docker container.
```
[BC-22-0-53722] PS C:\Run> $ips = Get-NetIPAddress | Where-Object { $_.AddressFamily -eq "IPv4" -and $_.IPAddress -ne "127.0.0.1" }

[BC-22-0-53722] PS C:\Run> $ips | ForEach-Object { Write-Output $_.IPAddress}
172.20.247.145
```

```
[BC-22-0-53722] PS C:\Run> $global:IP

IPAddress         : 127.0.0.1
InterfaceIndex    : 3
InterfaceAlias    : Loopback Pseudo-Interface 2
AddressFamily     : IPv4
Type              : Unicast
PrefixLength      : 8
PrefixOrigin      : WellKnown
SuffixOrigin      : WellKnown
AddressState      : Preferred
ValidLifetime     : Infinite ([TimeSpan]::MaxValue)
PreferredLifetime : Infinite ([TimeSpan]::MaxValue)
SkipAsSource      : False
PolicyStore       : ActiveStore

```

It appears that wherever $global:IP is being initialized has a problem.

I've updated to BcContainerHelper 4.0.16 and so far, Get-BcContainerIpAddress is returning a regular IP address.

Get-BcContainerIpAddress is failing again after working a few times.

I can further confirm that docker inspect fragment of code is working and it still seems to be a problem with $global:IP.

```
    $ip = Invoke-ScriptInBcContainer -containerName $containerName -scriptblock {
        if ($global:IP) {
            $ip = $global:IP
        }
        else {
            $ip = ""
            $ips = Get-NetIPAddress | Where-Object { $_.AddressFamily -eq "IPv4" -and $_.IPAddress -ne "127.0.0.1" }
            if ($ips) {
                $ips | ForEach-Object {
                    if ("$ip" -eq "") {
                        $ip = $_.IPAddress
                    }
                }
            }
            $global:IP = $ip
        }
        $ip
    }

Write-Output ("[" + $ip + "]")

$ip = ""

    if ("$ip" -eq "") {
        $inspect = docker inspect $containerName | ConvertFrom-Json
        $networks = $inspect.NetworkSettings.Networks
        $networks | get-member -MemberType NoteProperty | Select-Object Name | % {
            $name = $_.Name
            if (("$ip" -eq "") -and ("$networkName" -eq "" -or "$networkName" -eq "$name")) {
                $network = $networks | Select-Object -ExpandProperty $name
                $ip = $network.IPAddress
            }
        }
    }
    return $ip

```

The only way I get the Get-BcContainerIpAddress function to work consistently is to comment out the portion that runs inside the container and accesses $global:IP.

```
function Get-BcContainerIpAddress {
    [CmdletBinding()]
    Param (
        [string] $containerName = $bcContainerHelperConfig.defaultContainerName,
        [Parameter(Mandatory=$false)]
        [string] $networkName = ""
    )

$telemetryScope = InitTelemetryScope -name $MyInvocation.InvocationName -parameterValues $PSBoundParameters -includeParameters @()
try {

    $ip = Invoke-ScriptInBcContainer -containerName $containerName -scriptblock {
       #if ($global:IP) {
       #    $ip = $global:IP
       #}
       #else {
            $ip = ""
            $ips = Get-NetIPAddress | Where-Object { $_.AddressFamily -eq "IPv4" -and $_.IPAddress -ne "127.0.0.1" }
            if ($ips) {
                $ips | ForEach-Object {
                    if ("$ip" -eq "") {
                        $ip = $_.IPAddress
                    }
                }
            }
            $global:IP = $ip
       #}
        $ip
    }

    if ("$ip" -eq "") {
        $inspect = docker inspect $containerName | ConvertFrom-Json
        $networks = $inspect.NetworkSettings.Networks
        $networks | get-member -MemberType NoteProperty | Select-Object Name | % {
            $name = $_.Name
            if (("$ip" -eq "") -and ("$networkName" -eq "" -or "$networkName" -eq "$name")) {
                $network = $networks | Select-Object -ExpandProperty $name
                $ip = $network.IPAddress
            }
        }
    }
    return $ip
}
catch {
    TrackException -telemetryScope $telemetryScope -errorRecord $_
    throw
}
finally {
    TrackTrace -telemetryScope $telemetryScope
}
}

```

**Comment by freddydk** [comment]
Thanks
The $global:IP was created for performance optimizations, but it looks like it had side effects.
Will remove this.

**Comment by freddydk** [comment]
Strange that I haven't seen this on any machines elsewhere - what OS are you running on and what Docker SKU and version?

**Comment by freddydk** [comment]
I can see the OS is Windows Server 2019 (10.0.17763.3770 )

**Comment by freddydk** [comment]
Are you using special docker networks?

**Comment by BillJay1821** [comment]
No special docker networks.

**Comment by BillJay1821** [comment]
> $global:IP = "$ip"

I have reverted my other changes and applied the change you recommended as seen above and tested it on multiple computers with stopping and starting the docker container and Get-BcContainerIpAddress has been consistently returning a properly formed IP address.

**Comment by BillJay1821** [comment]
I just had an instance where the above change to only quote the $ip variable did NOT work and Get-BcContainerIpAddress returned an invalid IP address.

**Comment by BillJay1821** [comment]
I have been using the version of Get-BcContainerIpAddress seen below and still not having any issues.

```
function Get-BcContainerIpAddress {
    [CmdletBinding()]
    Param (
        [string] $containerName = $bcContainerHelperConfig.defaultContainerName,
        [Parameter(Mandatory=$false)]
        [string] $networkName = ""
    )

$telemetryScope = InitTelemetryScope -name $MyInvocation.InvocationName -parameterValues $PSBoundParameters -includeParameters @()
try {

    $ip = Invoke-ScriptInBcContainer -containerName $containerName -scriptblock {
        $ip = ""
        $ips = Get-NetIPAddress | Where-Object { $_.AddressFamily -eq "IPv4" -and $_.IPAddress -ne "127.0.0.1" }
        if ($ips) {
            $ips | ForEach-Object {
                if ("$ip" -eq "") {
                    $ip = $_.IPAddress
                }
            }
        }
       $ip
    }

    if ("$ip" -eq "") {
        $inspect = docker inspect $containerName | ConvertFrom-Json
        $networks = $inspect.NetworkSettings.Networks
        $networks | get-member -MemberType NoteProperty | Select-Object Name | % {
            $name = $_.Name
            if (("$ip" -eq "") -and ("$networkName" -eq "" -or "$networkName" -eq "$name")) {
                $network = $networks | Select-Object -ExpandProperty $name
                $ip = $network.IPAddress
            }
        }
    }
    return $ip
}
catch {
    TrackException -telemetryScope $telemetryScope -errorRecord $_
    throw
}
finally {
    TrackTrace -telemetryScope $telemetryScope
}
}
```

**Comment by freddydk** [comment]
In the above code  if $ip is empty string after the first check, the function will return two values.
You should remove the $ip line.

**Comment by BillJay1821** [comment]
@freddydk Is there a chance of getting the $global:IP change to Get-BcContainerIpAddress reverted?

**Comment by freddydk** [comment]
Yes, sorry - have been caught up in other things.
Next version will have the functionality reverted as you suggested.

**Comment by freddydk** [comment]
Available in preview now.

**Comment by freddydk** [comment]
Shipped in 5.0.0
